File: 3075-maximize-happiness-of-selected-children/3075-maximize-happiness-of-selected-children.java
class Solution {
    public long maximumHappinessSum(int[] happiness, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for(int i=0; i<happiness.length; i++) {
            pq.add(happiness[i]);
        }
        long ans = 0, cnt = 0;
        while(!pq.isEmpty() && k-- > 0) {
            int x = pq.poll();
            x -= cnt;
            if(x <= 0) return ans;
            ans += x;
            cnt++;
        }
        return ans;
    }
}

File: 3075-maximize-happiness-of-selected-children/3075-maximize-happiness-of-selected-children.cpp
class Solution {
public:
    long long maximumHappinessSum(vector<int>& happiness, int k) {
        long long ans = 0;
        priority_queue<int>pq;
        for(int x:happiness) {
            pq.push(x);
        }
        int cnt = 0;
        while(!pq.empty() && k--) {
            int cur = pq.top() - cnt; pq.pop();
            if(cur <= 0) return ans;
            ans += cur;
            cnt++;
        }
        return ans;
    }
};